### ISSUE TITLE
Allow arbitrary types in deserialization

### ISSUE THREAD

**Issue by adriangb** [wyfo/apischema#221]
I'd like to deserialize / validate w/ arbitrary types (similar to Pydantic's flag), but I can't figure out how to do so.

Here's what I've tried:

```python
from apischema import deserializer, deserialize

class MyClass:
    ...

@deserializer
def myclass_deser(item: MyClass) -> MyClass:
    return item

val = deserialize(MyClass, MyClass())
```

I get:

```
  File ".../apischema/deserialization/__init__.py", line 732, in method
    raise ValidationError([str(err)])
apischema.validation.errors.ValidationError: ValidationError(messages=['maximum recursion depth exceeded'], children={})
```

I would also expect that no conversion is needed in the first place (since `apischema` is not loading JSON like Pydantic does, it's starting from Python objects) if the types already match, but with no registered deserializer or conversion I get an `Unsupported` error from visitor.

**Comment by wyfo** [comment]
So, if I understand your use case, it's more about dynamic validation than validation.
That's indeed not the purpose of *apischema* (especially `apischema.deserialize`), and I don't really want to start mixing deserialization and dynamic type checking.

Maybe in a separate function; to be honest, in the first versions, dynamic type checking was implemented with `check_types` function but it has been removed since, because I was not convinced about its use, and because I wanted to enforce static type checking instead. And I don't know If validate recursively all the fields of your classes is needed here.

Maybe you could be interested by https://github.com/PrettyWood/typingx. Or, if you container arguments are simple types, something like this could work:
```python
from collections.abc import Collection, Sequence, Set
from typing import Any, get_args, get_origin

collection_constructors = {Collection: list, Sequence: list, Set: Set}

def validate_container(container: Any, objects: Collection) -> Collection:
    origin, args = get_origin(container), get_args(container)
    assert issubclass(origin, Collection) and all(isinstance(arg, type) for arg in args)
    if origin is tuple and args[1] is not ...:
        if len(args) != len(objects):
            raise ValueError(f"Expected {len(args)}-tuple, got {len(objects)} elements")
    else:
        assert len(args) == 1
        args = args * len(objects)
    for i, (tp, elt) in enumerate(zip(args, objects)):
        if not isinstance(elt, tp):
            raise ValueError(f"Element {i}: expected {tp}, found {elt.__class__}")
    return collection_constructors.get(origin, origin)(objects)
```

**Comment by adriangb** [comment]
> it's more about dynamic validation than validation

I don't see where dynamic vs. static comes into play here. I am parsing an object into a pre-defined type at runtime. That's different than runtime type checking, which is what typingx does. And I don't think it's any different than parsing a json-like object, except that non-json types are allowed. At least in Pydantic world, they're equivalent (`parse_obj`).

But I agree that if apischema, and `deserialize` in particular, is focused on deserialization, trying to parse objects with them doesn't make sense. I guess I got a bit confused because deserialize accepts an object not a serialized JSON string, and so I was thinking it'd be a bit more general purpose.

**Comment by wyfo** [comment]
Sorry for the delay of this response, I was focused on cythonization.

Actually, I was wrong. I've realized it while thinking about using *apischema* with MessagePack and its extensions/raw bytes. And indeed, MessagePack is not the only use case (yours is another one), so it is an important feature to have. Thank you to have bring it here.

I've also implemented a draft tonight: #239, feel free to look at it (last file of the PR is an example).

In this implementation, only non primitive classes, i.e. no `NewType`/list/int/str/`TypedDict`/etc., can be allowed (other can be passed, but are ignored). 

Validation is not well handled for the moment, but normally, type validators should not be run. In fact, if the type is already deserialized, it supposed to be valid (*apischema* could be called in another place). However, validators applied to the type at the upper level (for example in `Annotated`) must be run; that's the trickier part to implement (because currently type validators are merged with upper level ones before checking if the type is allowed).

Does it sound good to you?
